skrifa/src/outline/glyf/hint/engine/graphics_state.rs

```diff
@@ -0,0 +1,1130 @@
+    ///
+    /// Pops: n (value for loop Graphics State variable (integer))
+    ///
+    /// Pops a value, n, from the stack and sets the loop variable count to
+    /// that value. The loop variable works with the SHP\[a\], SHPIX[], IP[],
+    /// FLIPPT[], and ALIGNRP[]. The value n indicates the number of times
+    /// the instruction is to be repeated. After the instruction executes,
+    /// the loop variable is reset to 1.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#set-loop-variable>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L3287>
+    pub(super) fn op_sloop(&mut self) -> OpResult {
+        let n = self.value_stack.pop()?;
+        if n < 0 {
+            return Err(HintErrorKind::NegativeLoopCounter);
+        }
+        // As in FreeType, heuristically limit the number of loops to 16 bits.
+        self.graphics_state.loop_counter = (n as u32).min(0xFFFF);
+        Ok(())
+    }
+
+    /// Set minimum distance.
+    ///
+    /// SMD[] (0x1A)
+    ///
+    /// Pops: distance: value for minimum_distance (F26Dot6)
+    ///
+    /// Pops a value from the stack and sets the minimum_distance variable
+    /// to that value. The distance is assumed to be expressed in sixty-fourths
+    /// of a pixel.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#set-minimum_distance>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L4266>
+    pub(super) fn op_smd(&mut self) -> OpResult {
+        let distance = self.value_stack.pop()?;
+        self.graphics_state.min_distance = distance;
+        Ok(())
+    }
+
+    /// Instruction execution control.
+    ///
+    /// INSTCTRL[] (0x8E)
+    ///
+    /// Pops: s: selector flag (int32)
+    ///       value: used to set value of instruction_control (uint16 padded)
+    ///
+    /// Sets the instruction control state variable making it possible to turn
+    /// on or off the execution of instructions and to regulate use of
+    /// parameters set in the CVT program. INSTCTRL[ ] can only be executed in
+    /// the CVT program.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#instruction-execution-control>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L4871>
+    pub(super) fn op_instctrl(&mut self) -> OpResult {
+        let selector = self.value_stack.pop()? as u32;
+        let value = self.value_stack.pop()? as u32;
+        // Selectors are indices starting with 1; not flags.
+        // Convert index to flag.
+        let selector_flag = 1 << (selector - 1);
+        if !(1..=3).contains(&selector) || (value != 0 && value != selector_flag) {
+            return Ok(());
+        }
```

Comment:
what does it mean if the selector is out of this range? Is that an expected condition?

Reply:
Short answer is that we just ignore it.

The long answer is that FreeType has a "pedantic" mode that would fail here but I've never seen it enabled in practice because it breaks real fonts 🤷 

This is something I'd like to spend more time investigating in the future... what I'd like to see is a run of the hinter with full error checking and then fall back to an unhinted glyph if it fails. I've also seen real world font bugs where running FT in non-pedantic mode leads to actual visual artifacts. 

For now though, we stick with the status quo, for better or worse.